README
NAME

    URI::Find - Find URIs in arbitrary text

SYNOPSIS

      require URI::Find;
    
      my $finder = URI::Find->new(\&callback);
    
      $how_many_found = $finder->find(\$text);

DESCRIPTION

    This module does one thing: Finds URIs and URLs in plain text. It finds
    them quickly and it finds them all (or what URI.pm considers a URI to
    be.) It only finds URIs which include a scheme (http:// or the like),
    for something a bit less strict have a look at URI::Find::Schemeless.

    For a command-line interface, urifind is provided.

 Public Methods

    new

        my $finder = URI::Find->new(\&callback);

      Creates a new URI::Find object.

      &callback is a function which is called on each URI found. It is
      passed two arguments, the first is a URI object representing the URI
      found. The second is the original text of the URI found. The return
      value of the callback will replace the original URI in the text.

    find

        my $how_many_found = $finder->find(\$text);

      $text is a string to search and possibly modify with your callback.

      Alternatively, find can be called with a replacement function for the
      rest of the text:

        use CGI qw(escapeHTML);
        # ...
        my $how_many_found = $finder->find(\$text, \&escapeHTML);

      will not only call the callback function for every URL found (and
      perform the replacement instructions therein), but also run the rest
      of the text through escapeHTML(). This makes it easier to turn plain
      text which contains URLs into HTML (see example below).

 Protected Methods

    I got a bunch of mail from people asking if I'd add certain features to
    URI::Find. Most wanted the search to be less restrictive, do more
    heuristics, etc... Since many of the requests were contradictory, I'm
    letting people create their own custom subclasses to do what they want.

    The following are methods internal to URI::Find which a subclass can
    override to change the way URI::Find acts. They are only to be called
    inside a URI::Find subclass. Users of this module are NOT to use these
    methods.

    uri_re

        my $uri_re = $self->uri_re;

      Returns the regex for finding absolute, schemed URIs
      (http://www.foo.com and such). This, combined with
      schemeless_uri_re() is what finds candidate URIs.

      Usually this method does not have to be overridden.

    schemeless_uri_re

        my $schemeless_re = $self->schemeless_uri_re;

      Returns the regex for finding schemeless URIs (www.foo.com and such)
      and other things which might be URIs. By default this will match
      nothing (though it used to try to find schemeless URIs which started
      with www and ftp).

      Many people will want to override this method. See
      URI::Find::Schemeless for a subclass does a reasonable job of finding
      URIs which might be missing the scheme.

    uric_set

        my $uric_set = $self->uric_set;

      Returns a set matching the 'uric' set defined in RFC 2396 suitable
      for putting into a character set ([]) in a regex.

      You almost never have to override this.

    cruft_set

        my $cruft_set = $self->cruft_set;

      Returns a set of characters which are considered garbage. Used by
      decruft().

    decruft

        my $uri = $self->decruft($uri);

      Sometimes garbage characters like periods and parenthesis get
      accidentally matched along with the URI. In order for the URI to be
      properly identified, it must sometimes be "decrufted", the garbage
      characters stripped.

      This method takes a candidate URI and strips off any cruft it finds.

    recruft

        my $uri = $self->recruft($uri);

      This method puts back the cruft taken off with decruft(). This is
      necessary because the cruft is destructively removed from the string
      before invoking the user's callback, so it has to be put back
      afterwards.

    schemeless_to_schemed

        my $schemed_uri = $self->schemeless_to_schemed($schemeless_uri);

      This takes a schemeless URI and returns an absolute, schemed URI. The
      standard implementation supplies ftp:// for URIs which start with
      ftp., and http:// otherwise.

    is_schemed

        $obj->is_schemed($uri);

      Returns whether or not the given URI is schemed or schemeless. True
      for schemed, false for schemeless.

    badinvo

        __PACKAGE__->badinvo($extra_levels, $msg)

      This is used to complain about bogus subroutine/method invocations.
      The args are optional.

 Old Functions

    The old find_uri() function is still around and it works, but its
    deprecated.

EXAMPLES

    Store a list of all URIs (normalized) in the document.

      my @uris;
      my $finder = URI::Find->new(sub {
          my($uri) = shift;
          push @uris, $uri;
      });
      $finder->find(\$text);

    Print the original URI text found and the normalized representation.

      my $finder = URI::Find->new(sub {
          my($uri, $orig_uri) = @_;
          print "The text '$orig_uri' represents '$uri'\n";
          return $orig_uri;
      });
      $finder->find(\$text);

    Check each URI in document to see if it exists.

      use LWP::Simple;
    
      my $finder = URI::Find->new(sub {
          my($uri, $orig_uri) = @_;
          if( head $uri ) {
              print "$orig_uri is okay\n";
          }
          else {
              print "$orig_uri cannot be found\n";
          }
          return $orig_uri;
      });
      $finder->find(\$text);

    Turn plain text into HTML, with each URI found wrapped in an HTML
    anchor.

      use CGI qw(escapeHTML);
      use URI::Find;
    
      my $finder = URI::Find->new(sub {
          my($uri, $orig_uri) = @_;
          return qq|<a href="$uri">$orig_uri</a>|;
      });
      $finder->find(\$text, \&escapeHTML);
      print "<pre>$text</pre>";

NOTES

    Will not find URLs with Internationalized Domain Names or pretty much
    any non-ascii stuff in them. See
    http://rt.cpan.org/Ticket/Display.html?id=44226

AUTHOR

    Michael G Schwern <schwern@pobox.com> with insight from Uri Gutman,
    Greg Bacon, Jeff Pinyan, Roderick Schertler and others.

    Roderick Schertler <roderick@argon.org> maintained versions 0.11 to
    0.16.

    Darren Chamberlain wrote urifind.

LICENSE

    Copyright 2000, 2009-2010 by Michael G Schwern <schwern@pobox.com>.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See http://www.perlfoundation.org/artistic_license_1_0

SEE ALSO

    urifind, URI::Find::Schemeless, URI, RFC 3986 Appendix C



Changes
    Bug Fixes
	* URI first character matching [github 17] (lazarus)

20160806  Sat Aug  6 13:36:14 PDT 2016
    Distribution
    * Homepage is now metacpan
    * Now testing Strawberry Perl with AppVeyor
    * Kwalitee fix (Patrice Clement) [github 14]

20140709  Wed Jul  9 16:28:37 PDT 2014
    New Features
    * The "git" scheme is supported. (Schwern)
    * svn, ssh and svn+ssh schemes are supported. [rt.cpan.org 57490] (Schwern)
    * Added a --schemeless option to urifind. (Schwern)

    Bug Fixes
    * http:// is no longer matched [rt.cpan.org 63283] (Schwern)

    Backwards Incompatibilities
    * Previously, URIs stringified to their canonical version.  Now
      they stringify as written.  This results in less loss of
      information.  For example.  "Blah HTTP:://FOO.COM" previously
      would stringify as "http://foo.com/" and now it will stringify
      as "HTTP://FOO.COM".  To restore the old behavior you can call
      $uri->canonical. (Schwern)

    Distribution Changes
    * No longer using URI::URL. (Schwern)
    * Now requires URI 1.60 for Unicode support. (Schwern)


20140702  Wed Jul  2 13:41:47 PDT 2014
    New Features
    * IDNA (aka Unicode) domains are now supported. [github 3] (GwenDragon)
    * The list of TLDs for schemeless matching has been updated. [github 3] (GwenDragon)

    Bug Fixes
    * Handle balanced [], {} and quotes in addition to (). [rt.cpan.org 85053] (Schwern)
    * Don't mangle IPv6 URLs. [rt.cpan.org 85053] (Schwern)
    * Schemeless is more accurate about two letter TLDs. [github 3] (GwenDragon)

    Distribution Changes
    * Switched the issue tracker to Github. (Schwern)


20111103  Thu Nov  3 12:14:21 PDT 2011
    Bug Fixes
    * URI::URL::strict will no longer leak out of find() if the callback
      or filter fails. [rt.cpan.org 71153] (Carl Chambers)


20111020  Thu Oct 20 17:31:56 PDT 2011
    Bug Fixes
    * Things which look like URIs, but aren't, are now properly escaped like
      other text.  [rt.cpan.org 71658]

    New Features
    * Balanced parens in URIs are no longer stripped.  Example:
      "http://example.com/foo(bar)"  (Merten Falk)


20100505  Wed May  5 18:48:44 PDT 2010
    Test Fixes
    * Fixed t/urifind/find.t on Windows


20100504.1039  Tue May  4 10:39:23 PDT 2010
    Doc Fixes
    * Forgot to mention that we ship with urifind now.


20100504  Tue May  4 10:29:52 PDT 2010
    New Features
    * Added a urifind program. (Darren Chamberlain)
    Bug Fixes
    * The final semi-colon was being strippped form URLs found in HTML
      that ended with HTML entities. (Michael Peters)
      Example: http://google.com/search?q=&lt;html&gt;
    * URLs with leading dots, pluses and minuses are now found.
      [rt.cpan.org 57032]
      Example: stuff...http://example.com


20100211  Thu Feb 11 04:02:26 PST 2010
    Bug Fixes
    * Finding URIs inside brackets was pretty badly broken by
      the last release.  (Michael Peters)


20090319  Thu Mar 19 12:17:53 PDT 2009
    Bug Fixes
    * Schemeless now ignores the case of the TLD.

    New Features
    * Updated the list of accepted domains for finding schemeless URIs
      from the latest ICANN list.

    Docs
    * Add LICENSE section
    * Remove wildly out of date CAVEATS
    * Added an example of how to get a list of all URIs.
    * Updated INSTALL section to reflect new dependencies and Module::Build
      installation process
    * Regenerated the README file


20090316  Mon Mar 16 16:18:10 PDT 2009
    New Features
    * Added optional replacement function to find(). Now you 
      can not only replace URLs found, but also the rest of the text around
      them in one fell swoop. (Mike Schilli) [rt.cpan.org 20486]
    * Whitespace inside <...> is now ignored as per the suggestion of
      RFC 3986 appendix C. [rt.cpan.org 20483]

    Other
    * Michael G Schwern is now primary maintainer again.  Thanks for all your
      work, Roderick!
    * Repository moved to http://github.com/schwern/uri-find
    * Now requires Test::More
    * Verisoning scheme changed to ISO date integers
    * Minimum Perl version is now 5.6.0.


0.16 Fri Jul 22 06:00:24 EDT 2005
    - Oops, make the URI::Find::Schemeless->top_level_domain_re case
      insensitive, as it should be and the docs claimed it was.  Thanks
      to Todd Eigenschink.


0.15 Tue Mar 22 07:23:17 EST 2005
    - Have all functions croak if invoked with the wrong number of
      arguments.  Add URI::Find->badinvo.
        https://rt.cpan.org/NoAuth/Bug.html?id=1845
    - Mention DARREN's urifind script in the man page.
    - Oops, URI::URL::strict was turned on and left on.  Put it back the
      way you found it.  Thanks to Chris Nandor.
        https://rt.cpan.org/NoAuth/Bug.html?id=11906
    - Schemeless.pm:
        - Find '<x>intag.com</x>'.
        - Get $tldRe from a new class method, ->top_level_domain_re.
        - Update top level domain list.


0.14 Sat Oct  9 08:20:04 EDT 2004
    - Add copyright notice.
    - Add ] to main $cruftSet, } to schemeless $cruftSet, for
      [http://square.com] and {brace.com}.
    - quotemeta() $cruftSet.


0.13 Mon Jul  1 10:37:54 EDT 2002
    - Don't find any schemeless URIs with a plain URI::Find.  Previously
      it'd find ones which started with "ftp." and "www.", but it was
      more prone to false positives than URI::Find::Schemeless.
    - Have schemeless_to_schemed use http:// except in the specific case
      in which it uses ftp://.  Remove URI::Find::Schemeless's version.


0.12 Wed Mar 20 14:39:21 EST 2002
    - Improve the "wrap each URI found in an HTML anchor" example.
    - Release a new version so CPAN sees the maintainer change.


0.11 Thu Jul 26 14:43:49 EDT 2001
    - Michael passed the module to Roderick for maintenance.
    - Improve test suite.
    - Tweak URI::Find::Schemeless not to find Foo.p[ml].


0.10  Mon Jul 10 20:14:08 EDT 2000
    - Rearchitected the internals to allow simple subclassing
    - Added URI::Find::Schemeless (thanks Roderick)

lib/URI/Find.pm
package URI::Find;

require 5.006;

use strict;
use warnings;
use base qw(Exporter);
use vars qw($VERSION @EXPORT);

$VERSION        = 20160806;
@EXPORT         = qw(find_uris);

use constant YES => (1==1);
use constant NO  => !YES;

use Carp        qw(croak);

require URI;

my $reserved   = q(;/?:@&=+$,[]);
my $mark       = q(-_.!~*'());
my $unreserved = "A-Za-z0-9\Q$mark\E";
my $uric       = quotemeta($reserved) . '\p{isAlpha}' . $unreserved . "%";

# URI scheme pattern without the non-alpha numerics.
# Those are extremely uncommon and interfere with the match.
my($schemeRe) = qr/[a-zA-Z][a-zA-Z0-9\+]*/;
my($uricSet)  = $uric; # use new set

# Some schemes which URI.pm does not explicitly support.
my $extraSchemesRe = qr{^(?:git|svn|ssh|svn\+ssh)$};

# We need to avoid picking up 'HTTP::Request::Common' so we have a
# subset of uric without a colon ("I have no colon and yet I must poop")
my($uricCheat) = __PACKAGE__->uric_set;
$uricCheat =~ tr/://d;

# Identifying characters accidentally picked up with a URI.
my($cruftSet) = q{])\},.'";}; #'#


=head1 NAME

URI::Find - Find URIs in arbitrary text

=head1 SYNOPSIS

  require URI::Find;

  my $finder = URI::Find->new(\&callback);

  $how_many_found = $finder->find(\$text);

=head1 DESCRIPTION

This module does one thing: Finds URIs and URLs in plain text.  It
finds them quickly and it finds them B<all> (or what URI.pm considers
a URI to be.)  It only finds URIs which include a scheme (http:// or
the like), for something a bit less strict have a look at
L<URI::Find::Schemeless|URI::Find::Schemeless>.

For a command-line interface, L<urifind> is provided.

=head2 Public Methods

=over 4

=item B<new>

  my $finder = URI::Find->new(\&callback);

Creates a new URI::Find object.

&callback is a function which is called on each URI found.  It is
passed two arguments, the first is a URI object representing the URI
found.  The second is the original text of the URI found.  The return
value of the callback will replace the original URI in the text.

=cut

sub new {
    @_ == 2 || __PACKAGE__->badinvo;
    my($proto, $callback) = @_;
    my($class) = ref $proto || $proto;
    my $self = bless {}, $class;

    $self->{callback} = $callback;

    return $self;
}

=item B<find>

  my $how_many_found = $finder->find(\$text);

$text is a string to search and possibly modify with your callback.

Alternatively, C<find> can be called with a replacement function for
the rest of the text:

  use CGI qw(escapeHTML);
  # ...
  my $how_many_found = $finder->find(\$text, \&escapeHTML);

will not only call the callback function for every URL found (and
perform the replacement instructions therein), but also run the rest
of the text through C<escapeHTML()>. This makes it easier to turn
plain text which contains URLs into HTML (see example below).

=cut

sub find {
    @_ == 2 || @_ == 3 || __PACKAGE__->badinvo;
    my($self, $r_text, $escape_func) = @_;

    # Might be slower, but it makes the code simpler
    $escape_func ||= sub { return $_[0] };

    # Store the escape func in the object temporarily for use
    # by other methods.
    local $self->{escape_func} = $escape_func;

    $self->{_uris_found} = 0;

    # Yes, evil.  Basically, look for something vaguely resembling a URL,
    # then hand it off to URI for examination.  If it passes, throw
    # it to a callback and put the result in its place.
    local $SIG{__DIE__} = 'DEFAULT';
    my $uri_cand;
    my $uri;

    my $uriRe = sprintf '(?:%s|%s)', $self->uri_re, $self->schemeless_uri_re;

    $$r_text =~ s{ (.*?) (?:(<(?:URL:)?)(.+?)(>)|($uriRe)) | (.+?)$ }{
        my $replace = '';
        if( defined $6 ) {
            $replace = $escape_func->($6);
        }
        else {
            my $maybe_uri = '';

            $replace = $escape_func->($1) if length $1;

            if( defined $2 ) {
                $maybe_uri = $3;
                my $is_uri = do {  # Don't alter $1...
                    $maybe_uri =~ s/\s+//g;
                    $maybe_uri =~ /^$uriRe/;
                };

                if( $is_uri ) {
                    $replace .= $escape_func->($2);
                    $replace .= $self->_uri_filter($maybe_uri);
                    $replace .= $escape_func->($4);
                }
                else {
                    # the whole text inside of the <...> was not a url, but
                    # maybe it has a url (like an HTML <a> link)
                    my $has_uri = do { # Don't alter $1...
                        $maybe_uri = $3;
                        $maybe_uri =~ /$uriRe/;
                    };
                    if( $has_uri ) {
                        my $pre = $2;
                        my $post = $4;
                        do { $self->find(\$maybe_uri, $escape_func) };
                        $replace .= $escape_func->($pre);
                        $replace .= $maybe_uri;  # already escaped by find()
                        $replace .= $escape_func->($post);
                    }
                    else {
                        $replace .= $escape_func->($2.$3.$4);
                    }
                }
            }
            else {
                $replace .= $self->_uri_filter($5);
            }
        }

        $replace;
    }gsex;

    return $self->{_uris_found};
}


sub _uri_filter {
    my($self, $orig_match) = @_;

    # A heuristic.  Often you'll see things like:
    # "I saw this site, http://www.foo.com, and its really neat!"
    # or "Foo Industries (at http://www.foo.com)"
    # We want to avoid picking up the trailing paren, period or comma.
    # Of course, this might wreck a perfectly valid URI, more often than
    # not it corrects a parse mistake.
    $orig_match = $self->decruft($orig_match);

    my $replacement = '';
    if( my $uri = $self->_is_uri(\$orig_match) ) {
        # It's a URI
        $self->{_uris_found}++;
        $replacement = $self->{callback}->($uri, $orig_match);
    }
    else {
        # False alarm
        $replacement = $self->{escape_func}->($orig_match);
    }

    # Return recrufted replacement
    return $self->recruft($replacement);
}


=back

=head2 Protected Methods

I got a bunch of mail from people asking if I'd add certain features
to URI::Find.  Most wanted the search to be less restrictive, do more
heuristics, etc...  Since many of the requests were contradictory, I'm
letting people create their own custom subclasses to do what they
want.

The following are methods internal to URI::Find which a subclass can
override to change the way URI::Find acts.  They are only to be called
B<inside> a URI::Find subclass.  Users of this module are NOT to use
these methods.

=over

=item B<uri_re>

  my $uri_re = $self->uri_re;

Returns the regex for finding absolute, schemed URIs
(http://www.foo.com and such).  This, combined with
schemeless_uri_re() is what finds candidate URIs.

Usually this method does not have to be overridden.

=cut

sub uri_re {
    @_ == 1 || __PACKAGE__->badinvo;
    my($self) = shift;
    return sprintf '%s:[%s][%s#]*', $schemeRe,
                                    $uricCheat,
                                    $self->uric_set;
}

=item B<schemeless_uri_re>

  my $schemeless_re = $self->schemeless_uri_re;

Returns the regex for finding schemeless URIs (www.foo.com and such) and
other things which might be URIs.  By default this will match nothing
(though it used to try to find schemeless URIs which started with C<www>
and C<ftp>).

Many people will want to override this method.  See L<URI::Find::Schemeless>
for a subclass does a reasonable job of finding URIs which might be missing
the scheme.

=cut

sub schemeless_uri_re {
    @_ == 1 || __PACKAGE__->badinvo;
    my($self) = shift;
    return qr/\b\B/; # match nothing
}

=item B<uric_set>

  my $uric_set = $self->uric_set;

Returns a set matching the 'uric' set defined in RFC 2396 suitable for
putting into a character set ([]) in a regex.

You almost never have to override this.

=cut

sub uric_set {
    @_ == 1 || __PACKAGE__->badinvo;
    return $uricSet;
}

=item B<cruft_set>

  my $cruft_set = $self->cruft_set;

Returns a set of characters which are considered garbage.  Used by
decruft().

=cut

sub cruft_set {
    @_ == 1 || __PACKAGE__->badinvo;
    return $cruftSet;
}

=item B<decruft>

  my $uri = $self->decruft($uri);

Sometimes garbage characters like periods and parenthesis get
accidentally matched along with the URI.  In order for the URI to be
properly identified, it must sometimes be "decrufted", the garbage
characters stripped.

This method takes a candidate URI and strips off any cruft it finds.

=cut

my %balanced_cruft = (
    '('         => ')',
    '{'         => '}',
    '['         => ']',
    '"'         => '"',
    q[']        => q['],
);

sub decruft {
    @_ == 2 || __PACKAGE__->badinvo;
    my($self, $orig_match) = @_;

    $self->{start_cruft} = '';
    $self->{end_cruft} = '';

    if( $orig_match =~ s/([\Q$cruftSet\E]+)$// ) {
        # urls can end with HTML entities if found in HTML so let's put back semicolons
        # if this looks like the case
        my $cruft = my $orig_cruft = $1;

        if( $cruft =~ /^;/ && $orig_match =~ /\&(\#[1-9]\d{1,3}|[a-zA-Z]{2,8})$/) {
            $orig_match .= ';';
            $cruft =~ s/^;//;
        }

        my $compare = sub { index( $orig_cruft, $_[ 1 ] )
                        <=> index( $orig_cruft, $_[ 0 ] ) };

        for my $open (sort { $compare->( $a, $b ) } keys %balanced_cruft) {
            my $close = $balanced_cruft{ $open };

            $self->recruft_balanced(\$orig_match, \$cruft, $open, $close);
        }

        $self->{end_cruft} = $cruft if $cruft;
    }

    return $orig_match;
}


sub recruft_balanced {
    my $self = shift;
    my($orig_match, $cruft, $open, $close) = @_;

    my $open_count  = () = $$orig_match =~ m{\Q$open}g;
    my $close_count = () = $$orig_match =~ m{\Q$close}g;

    $open eq $close and $open_count and $open_count == $close_count
        and $open_count % 2 > 0 and $close_count--;

    if ($$cruft =~ m{ \A \Q$close\E }mx && $open_count == ($close_count + 1) ) {
        $$orig_match .= $close;
        $$cruft =~ s{ \A \Q$close\E }{}mx;
    }

    return;
}


=item B<recruft>

  my $uri = $self->recruft($uri);

This method puts back the cruft taken off with decruft().  This is necessary
because the cruft is destructively removed from the string before invoking
the user's callback, so it has to be put back afterwards.

=cut

#'#

sub recruft {
    @_ == 2 || __PACKAGE__->badinvo;
    my($self, $uri) = @_;

    return $self->{start_cruft} . $uri . $self->{end_cruft};
}

=item B<schemeless_to_schemed>

  my $schemed_uri = $self->schemeless_to_schemed($schemeless_uri);

This takes a schemeless URI and returns an absolute, schemed URI.  The
standard implementation supplies ftp:// for URIs which start with ftp.,
and http:// otherwise.

=cut

sub schemeless_to_schemed {
    @_ == 2 || __PACKAGE__->badinvo;
    my($self, $uri_cand) = @_;

    $uri_cand =~ s|^(<?)ftp\.|$1ftp://ftp\.|
        or $uri_cand =~ s|^(<?)|${1}http://|;

    return $uri_cand;
}

=item B<is_schemed>

  $obj->is_schemed($uri);

Returns whether or not the given URI is schemed or schemeless.  True for
schemed, false for schemeless.

=cut

sub is_schemed {
    @_ == 2 || __PACKAGE__->badinvo;
    my($self, $uri) = @_;
    return scalar $uri =~ /^<?$schemeRe:/;
}

=item I<badinvo>

  __PACKAGE__->badinvo($extra_levels, $msg)

This is used to complain about bogus subroutine/method invocations.
The args are optional.

=cut

sub badinvo {
    my $package = shift;
    my $level   = @_ ? shift : 0;
    my $msg     = @_ ? " (" . shift() . ")" : '';
    my $subname = (caller $level + 1)[3];
    croak "Bogus invocation of $subname$msg";
}

=back

=head2 Old Functions

The old find_uri() function is still around and it works, but its
deprecated.

=cut

# Old interface.
sub find_uris (\$&) {
    @_ == 2 || __PACKAGE__->badinvo;
    my($r_text, $callback) = @_;

    my $self = __PACKAGE__->new($callback);
    return $self->find($r_text);
}


=head1 EXAMPLES

Store a list of all URIs (normalized) in the document.

  my @uris;
  my $finder = URI::Find->new(sub {
      my($uri) = shift;
      push @uris, $uri;
  });
  $finder->find(\$text);

Print the original URI text found and the normalized representation.

  my $finder = URI::Find->new(sub {
      my($uri, $orig_uri) = @_;
      print "The text '$orig_uri' represents '$uri'\n";
      return $orig_uri;
  });
  $finder->find(\$text);

Check each URI in document to see if it exists.

  use LWP::Simple;

  my $finder = URI::Find->new(sub {
      my($uri, $orig_uri) = @_;
      if( head $uri ) {
          print "$orig_uri is okay\n";
      }
      else {
          print "$orig_uri cannot be found\n";
      }
      return $orig_uri;
  });
  $finder->find(\$text);


Turn plain text into HTML, with each URI found wrapped in an HTML anchor.

  use CGI qw(escapeHTML);
  use URI::Find;

  my $finder = URI::Find->new(sub {
      my($uri, $orig_uri) = @_;
      return qq|<a href="$uri">$orig_uri</a>|;
  });
  $finder->find(\$text, \&escapeHTML);
  print "<pre>$text</pre>";

=cut


sub _is_uri {
    @_ == 2 || __PACKAGE__->badinvo;
    my($self, $r_uri_cand) = @_;

    my $uri = $$r_uri_cand;

    # Translate schemeless to schemed if necessary.
    $uri = $self->schemeless_to_schemed($uri) if
      $uri =~ $self->schemeless_uri_re   and
      $uri !~ /^<?$schemeRe:/;

    eval {
        $uri = URI->new($uri);

        # Throw out anything with an invalid scheme.
        my $has_invalid_scheme = $uri->isa("URI::_foreign") &&
                                 $uri->scheme !~ $extraSchemesRe;

        # Toss out things like http:// but keep file:///
        my $is_empty = $uri =~ m{^$schemeRe://$};

        undef $uri if $has_invalid_scheme || $is_empty;
    };

    if($@ || !defined $uri) {   # leave everything untouched, its not a URI.
        return NO;
    }
    else {                      # Its a URI.
        return $uri;
    }
}


=head1 NOTES

Will not find URLs with Internationalized Domain Names or pretty much
any non-ascii stuff in them.  See
L<http://rt.cpan.org/Ticket/Display.html?id=44226>


=head1 AUTHOR

Michael G Schwern <schwern@pobox.com> with insight from Uri Gutman,
Greg Bacon, Jeff Pinyan, Roderick Schertler and others.

Roderick Schertler <roderick@argon.org> maintained versions 0.11 to 0.16.

Darren Chamberlain wrote urifind.


=head1 LICENSE

Copyright 2000, 2009-2010, 2014, 2016 by Michael G Schwern E<lt>schwern@pobox.comE<gt>.

This program is free software; you can redistribute it and/or 
modify it under the same terms as Perl itself.

See F<http://www.perlfoundation.org/artistic_license_1_0>

=head1 SEE ALSO

L<urifind>, L<URI::Find::Schemeless>, L<URI>, RFC 3986 Appendix C

=cut

1;
